### PR TITLE
ref(new-widget-builder-experience): Add debounce widget prop to the visualisation component - [DD-758]

### PR DESCRIPTION
**Pull request by priscilawebdev**
debounce the prop `widget` in the visualization component so then a request to the `eventsV2` or `metrics/data` API is not always fired when the user types a new character in the alias field 

~TODO:~
~have to find out how to test this, but will do it in a separate task~

we have` jest.mock('lodash/debounce', () => jest.fn(fn => fn));`in the `setup.ts` and unmock it was necessary

### CODE REVIEW

File: static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx [hunk 1]
```diff
@@ -34,9 +39,29 @@ export function VisualizationStep({
   widgetBuilderNewDesign,
   widget,
 }: Props) {
-  const options = widgetBuilderNewDesign
-    ? Object.keys(displayTypes).filter(key => key !== DisplayType.TOP_N)
-    : Object.keys(displayTypes);
+  const [debouncedWidget, setDebouncedWidget] = useState(widget);
+  const previousWidget = usePrevious(widget);
+
+  const debouceWidget = useCallback(
```
Comment (JonasBa) [inline static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx:45]: Small typo
```suggestion
  const debounceWidget = useCallback(
```

File: static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx [hunk 2]
```diff
@@ -34,9 +39,29 @@ export function VisualizationStep({
...
+      debouceWidget(widget);
+    }
+  }, [widget, previousWidget]);
+
+  const options = useMemo(
+    () =>
+      widgetBuilderNewDesign
+        ? Object.keys(displayTypes).filter(key => key !== DisplayType.TOP_N)
+        : Object.keys(displayTypes),
+    []
```
Comment (JonasBa) [inline static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx:63]: You need dependencies here

File: static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx [hunk 3]
```diff
@@ -34,9 +39,29 @@ export function VisualizationStep({
...
+  const debounceWidget = useCallback(
+    debounce((value: Widget) => {
+      setDebouncedWidget(value);
+    }, DEFAULT_DEBOUNCE_DURATION),
+    []
+  );
+
+  useEffect(() => {
+    if (!isEqual(previousWidget, widget)) {
+      debouceWidget(widget);
```
Comment (priscilawebdev) [inline static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx:54]:
```suggestion
      debounceWidget(widget);
```

File: static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx [hunk 4]
```diff
@@ -34,9 +39,29 @@ export function VisualizationStep({
...
+      debounceWidget(widget);
+    }
+  }, [widget, previousWidget]);
+
+  const options = useMemo(
+    () =>
+      widgetBuilderNewDesign
+        ? Object.keys(displayTypes).filter(key => key !== DisplayType.TOP_N)
+        : Object.keys(displayTypes),
+    []
```
Comment (priscilawebdev) [inline static/app/views/dashboardsV2/widgetBuilder/buildSteps/visualizationStep.tsx:63]:
```suggestion
    [widgetBuilderNewDesign]
```

File: static/app/views/dashboardsV2/widgetCard/widgetCardChartContainer.tsx [hunk 5]
```diff
@@ -70,55 +54,7 @@ export function WidgetCardChartContainer({
   expandNumbers,
   onDataFetched,
 }: Props) {
-  function issueTableResultComponent({
```
Comment (priscilawebdev) [inline static/app/views/dashboardsV2/widgetCard/widgetCardChartContainer.tsx:73]: took the opportunity that I was checking for re-renders and moved this to a new component called issueWidgetCard.

**Review by narsaynorath — APPROVED** [review]
Just one comment about the filtering for TOP_N, but this is a neat way to debounce the widget preview

**Pull request merged**